File: ikev2/src/state.rs
use bytes::Bytes;
use std::{collections::VecDeque, time::Instant};

use super::{AssignedConfig, ChildSaInstall};

#[derive(Clone, Copy, Default)]
pub enum IkeSaState {
    #[default]
    Stopped,
    Starting,
    IkeSaInitSent,
    IkeSaInitEstablished,
    IkeAuthBootstrap,
    IkeAuthEapInProgress,
    ChildSaInstalling,
    Running,
    Failed,
}

impl IkeSaState {
    pub fn is_terminal(self) -> bool {
        matches!(self, Self::Stopped | Self::Running | Self::Failed)
    }

    pub fn is_handshaking(self) -> bool {
        matches!(
            self,
            Self::Starting
                | Self::IkeSaInitSent
                | Self::IkeSaInitEstablished
                | Self::IkeAuthBootstrap
                | Self::IkeAuthEapInProgress
                | Self::ChildSaInstalling
        )
    }
}

#[derive(Default)]
pub struct IkeSa {
    pub state: IkeSaState,
    pub next_request_message_id: u32,
    pub expected_response_message_id: Option<u32>,
    pub last_completed_response_message_id: Option<u32>,
    pub first_ike_auth_seen: bool,
    pub initiator_spi: u64,
    pub responder_spi: u64,
    pub sa_init_request: Option<Bytes>,
    pub sa_init_response: Option<Bytes>,
    pub sa_init_cookie: Option<Box<[u8]>>,
    pub initiator_nonce: Option<Bytes>,
    pub responder_nonce: Option<Bytes>,
    pub initiator_ke: Option<KeyExchangePayload>,
    pub responder_ke: Option<KeyExchangePayload>,
    pub key_material: Option<IkeKeyMaterial>,
    pub nat: NatDetectionState,
    pub auth: IkeAuthState,
    pub assigned_config: Option<AssignedConfig>,
    pub negotiating_child_sa: Option<NegotiatingChildSa>,
    pub active_child_sa: Option<ActiveChildSa>,
    pub outbound_request: Option<OutboundRequest>,
    pub inbound_request_history: VecDeque<InboundRequestResponse>,
    pub peer: PeerCapabilities,
    pub inbound_fragments: Option<InboundFragmentReassembly>,
    pub failure_reason: Option<String>,
    pub suppress_error_auth_failed_notify: bool,
}

impl IkeSa {
    pub fn new() -> Self {
        Default::default()
    }

    pub fn reset(&mut self) {
        *self = Self::new();
    }

    pub fn set_outbound_request(&mut self, message_id: u32, packets: Vec<Bytes>) {
        self.outbound_request = Some(OutboundRequest { message_id, packets });
    }

    pub fn clear_outbound_request(&mut self) {
        self.outbound_request = None;
    }
}

#[derive(Clone)]
pub struct KeyExchangePayload {
    pub dh_group: u16,
    pub value: Box<[u8]>,
}

#[derive(Clone)]
pub struct IkeKeyMaterial {
    pub sk_d: Box<[u8]>,
    pub sk_ai: Box<[u8]>,
    pub sk_ar: Box<[u8]>,
    pub sk_ei: Box<[u8]>,
    pub sk_er: Box<[u8]>,
    pub sk_pi: Box<[u8]>,
    pub sk_pr: Box<[u8]>,
}

#[derive(Default)]
pub struct NatDetectionState {
    pub source_hash_local: Option<Box<[u8]>>,
    pub destination_hash_local: Option<Box<[u8]>>,
    pub source_hash_remote: Option<Box<[u8]>>,
    pub destination_hash_remote: Option<Box<[u8]>>,
    pub detected: bool,
}

impl NatDetectionState {
    pub fn new() -> Self {
        Self {
            source_hash_local: None,
            destination_hash_local: None,
            source_hash_remote: None,
            destination_hash_remote: None,
            detected: false,
        }
    }
}

#[derive(Default)]
pub struct IkeAuthState {
    pub first_idi_payload: Option<Bytes>,
    pub peer_idr_payload: Option<Bytes>,
    pub peer_auth_method: Option<u8>,
    pub peer_signature_hash_algorithms: Box<[u16]>,
    pub peer_cert_der: Option<Bytes>,
    pub peer_cert_chain: Vec<Bytes>,
    pub local_eap_msk: Option<Box<[u8]>>,
}

impl IkeAuthState {
    pub fn new() -> Self {
        Self {
            first_idi_payload: None,
            peer_idr_payload: None,
            peer_auth_method: None,
            peer_signature_hash_algorithms: Box::new([]),
            peer_cert_der: None,
            peer_cert_chain: Vec::new(),
            local_eap_msk: None,
        }
    }
}

pub struct NegotiatingChildSa {
    pub inbound_spi: u32,
}

pub struct ActiveChildSa {
    pub install: ChildSaInstall,
}

pub struct OutboundRequest {
    pub message_id: u32,
    pub packets: Vec<Bytes>,
}

pub struct InboundRequestResponse {
    pub message_id: u32,
    pub response_packets: Vec<Bytes>,
}

pub struct WireCheckpoint {
    pub message_id: u32,
    pub packet: Bytes,
}

#[derive(Default)]
pub struct SaInitState {
    pub request: Option<WireCheckpoint>,
    pub response: Option<WireCheckpoint>,
    pub cookie: Option<Box<[u8]>>,
}

impl SaInitState {
    pub fn new() -> Self {
        Default::default()
    }
}

#[derive(Default)]
pub struct EapRoundState {
    pub inbound_request: Option<WireCheckpoint>,
    pub outbound_response: Option<WireCheckpoint>,
    pub round_index: u16,
}

impl EapRoundState {
    pub fn new() -> Self {
        Default::default()
    }
}

#[derive(Default)]
pub struct PeerCapabilities {
    pub supports_eap_only_authentication: bool,
    pub supports_message_id_sync: bool,
    pub supports_fragmentation: bool,
}

impl PeerCapabilities {
    pub fn new() -> Self {
        Default::default()
    }
}

pub struct InboundFragmentReassembly {
    pub exchange_type: u8,
    pub message_id: u32,
    pub first_inner_payload: Option<u8>,
    pub fragments: Vec<Option<Bytes>>,
    pub expires_at: Instant,
}

File: control/state.go
package control

import (
	"net"
	"time"

	"swan/transport"
	"swan/wire"
	"swan/wire/payload"
	"swan/xcrypto"
)

// Phase is the control-plane session phase. Phases advance strictly forward:
//
//	Stopped -> Starting -> SAInitSent -> SAInitEstablished -> AuthBootstrap
//	  -> AuthEAPInProgress -> ChildInstalling -> Running
//	any phase                 -> Failed
type Phase uint8

const (
	PhaseStopped Phase = iota
	PhaseStarting
	PhaseSAInitSent
	PhaseSAInitEstablished
	PhaseAuthBootstrap
	PhaseAuthEAPInProgress
	PhaseChildInstalling
	PhaseRunning
	PhaseFailed
)

// Terminal reports whether no further phase transitions may occur without a
// Reset/Close.
func (p Phase) Terminal() bool {
	return p == PhaseStopped || p == PhaseRunning || p == PhaseFailed
}

// State is the live IKE SA state, owned by exactly one worker at a time
// (Handshake during the handshake, Running afterwards). Nil slices mean
// "not yet received". Checkpoint copies (SAInitRequest/Response, Cookie,
// nonces, certs) own their bytes; everything parsed elsewhere aliases input
// buffers only for the duration of the parse.
type State struct {
	Phase Phase

	// Message-id bookkeeping (RFC 7296 windowing, initiator view).
	NextRequestMessageID           uint32
	ExpectedResponseMessageID      uint32
	HasExpectedResponse            bool
	LastCompletedResponseMessageID uint32
	HasLastCompletedResponse       bool
	FirstIKEAuthSeen               bool

	InitiatorSPI uint64
	ResponderSPI uint64

	// SA_INIT checkpoints: full wire packets are kept because AUTH signing
	// needs the exact bytes.
	SAInitRequest  []byte
	SAInitResponse []byte
	Cookie         []byte

	InitiatorNonce []byte
	ResponderNonce []byte

	InitiatorKE payload.KeyExchange
	ResponderKE payload.KeyExchange
	LocalDH     *xcrypto.DHKey

	IKEKeys     *xcrypto.IKEKeys
	SelectedIKE *xcrypto.Selection
	SelectedESP *xcrypto.Selection

	NAT  NatState
	Auth AuthState

	NegotiatingChild *NegotiatingChildSA
	ActiveChild      *ChildSA

	// OutboundRequest is the retransmission checkpoint of the current
	// in-flight request (built frames kept for identical resends).
	OutboundRequest *Checkpoint
	// InboundHistory caches responses to peer requests (bounded, delete
	// replay support).
	InboundHistory []CachedResponse

	Peer             PeerCapabilities
	InboundFragments *FragmentReassembly

	Assigned      *AssignedConfig
	FailureReason string
	// SuppressAuthFailedNotify avoids echoing AUTHENTICATION_FAILED back at
	// the peer (swan2 behavior).
	SuppressAuthFailedNotify bool
}

// NatState tracks the four NAT-D hashes plus the detected verdict.
type NatState struct {
	SourceHashLocal       []byte
	DestinationHashLocal  []byte
	SourceHashRemote      []byte
	DestinationHashRemote []byte
	Detected              bool
}

// AuthState aggregates everything learned from IKE_AUTH responses.
type AuthState struct {
	FirstIDiPayload             []byte // IDi body sent in bootstrap AUTH
	PeerIDr                     []byte // latest IDr body
	PeerAuthMethod              uint8
	PeerSignatureHashAlgorithms []uint16
	PeerCerts                   [][]byte // X.509 DER chain (CERT payloads)
	LocalEAPMSK                 []byte
}

// NegotiatingChildSA holds the initiator-side CHILD_SA before the final
// IKE_AUTH confirms it.
type NegotiatingChildSA struct {
	InboundSPI uint32
}

// ChildSA is the active CHILD_SA identity: SPIs plus raw TSi/TSr bodies
// (for the installer/logging surface).
type ChildSA struct {
	InboundSPI  uint32
	OutboundSPI uint32
	TSi         []byte
	TSr         []byte
}

// AssignedConfig is the CP reply content (internal address + DNS).
type AssignedConfig struct {
	InternalIPv4       net.IP
	InternalIPv6       net.IP
	InternalIPv6Prefix uint8
	DNS4               []net.IP
	DNS6               []net.IP
	// AddressExpirySeconds is the CP lease lifetime from
	// INTERNAL_ADDRESS_EXPIRY, zero when the responder did not send one.
	// It is reported for the caller; renewing the lease is out of scope
	// for this implementation.
	AddressExpirySeconds uint32
}

// Checkpoint is an outbound request kept for retransmission.
type Checkpoint struct {
	MessageID uint32
	Packets   []*transport.Frame
}

// CachedResponse replays a previously built response for a duplicate peer
// request (bounded history; delete exchanges re-send the same bytes).
type CachedResponse struct {
	MessageID       uint32
	ResponsePackets []*transport.Frame
}

// PeerCapabilities aggregates negotiated NOTIFY capabilities.
type PeerCapabilities struct {
	SupportsEAPOnlyAuthentication bool
	SupportsMessageIDSync         bool
	SupportsFragmentation         bool
}

// FragmentReassembly accumulates inbound SKF fragments for one message.
type FragmentReassembly struct {
	ExchangeType      wire.ExchangeType
	MessageID         uint32
	FirstInnerPayload wire.PayloadType
	Fragments         [][]byte
	ExpiresAt         time.Time
}

// NewState returns the zero state in PhaseStopped.
func NewState() *State {
	return &State{Phase: PhaseStopped}
}

// Reset returns the state to zero (used at the start of a new handshake and
// during cleanup).
func (s *State) Reset() {
	*s = *NewState()
}

// ClearSession wipes key material, nonces, child state, checkpoints and
// fragment state — the cleanup half of a graceful close.
func (s *State) ClearSession() {
	s.IKEKeys = nil
	s.InitiatorNonce = nil
	s.ResponderNonce = nil
	s.Cookie = nil
	s.NegotiatingChild = nil
	s.ActiveChild = nil
	s.OutboundRequest = nil
	s.InboundHistory = s.InboundHistory[:0]
	s.InboundFragments = nil
	s.Assigned = nil
	s.LocalDH = nil
	s.Auth.LocalEAPMSK = nil
	s.Auth.FirstIDiPayload = nil
	s.Auth.PeerIDr = nil
	s.Auth.PeerSignatureHashAlgorithms = nil
	s.Auth.PeerCerts = nil
	s.SAInitRequest = nil
	s.SAInitResponse = nil
	s.HasExpectedResponse = false
	s.HasLastCompletedResponse = false
	s.FirstIKEAuthSeen = false
	s.SuppressAuthFailedNotify = false
}
